bin/propolis-server/src/lib/initializer.rs

```diff
@@ -564,18 +568,59 @@ impl<'a> MachineInitializer<'a> {
                     self.devices
                         .insert(format!("pci-nvme-{bdf}"), nvme.clone());
                     block::attach(nvme.clone(), backend).unwrap();
-                    chipset.pci_attach(bdf, nvme);
+                    chipset.pci_attach(bdf, nvme.clone());
+                    nvme
                 }
             };
 
-            if let Some((id, backend)) = crucible {
-                let prev = self.crucible_backends.insert(id, backend);
+            if let Some((disk_id, backend)) = crucible {
+                let block_size = backend.block_size().await;
+                let prev = self.crucible_backends.insert(disk_id, backend);
                 if prev.is_some() {
                     return Err(Error::new(
                         ErrorKind::InvalidInput,
-                        format!("multiple disks with id {}", id),
+                        format!("multiple disks with id {}", disk_id),
                     ));
                 }
+
+                let Some(block_size) = block_size else {
+                    slog::error!(
+                        self.log,
+                        "Could not get Crucible backend block size, \
+                        virtual disk metrics can't be reported for it";
+                        "disk_id" => %disk_id,
+                    );
+                    return Ok(());
+                };
+
+                // Register the block device as a metric producer, if we've been
+                // setup to do so. Note we currently only do this for the Crucible
+                // backend, in which case we have the disk ID.
+                if let Some(registry) = &self.producer_registry {
+                    let stats = VirtualDiskProducer::new(
+                        block_size,
+                        self.properties.id,
+                        disk_id,
+                        &self.properties.metadata,
+                    );
+                    if let Err(e) = registry.register_producer(stats.clone()) {
+                        slog::error!(
+                            self.log,
+                            "Could not register virtual disk producer, \
+                            metrics will not be produced";
+                            "disk_id" => %disk_id,
+                            "error" => ?e,
+                        );
```

Comment:
If we can't register stats, will this bail out of the loop early?

It seems like we are in the loop:
`for (name, device_spec) in &self.spec.devices.storage_devices {`

And, if we `retrun Ok`, will the caller even know that not all of their devices were not handled.

The same is true about if we can't get a block size above, we could bail early and not attach other disks.

Reply:
Nice catch! This is definitely wrong, we should be `continue`ing here instead. I'll update that, thanks.

Reply:
Switched to a `continue` in 15327f2